### DISCUSSION TITLE
Intelligent scanning of channels

### DISCUSSION THREAD

**Discussion by BellezaEmporium** [iptv-org/community#62] [Ideas]
## Summary
We have, as of today, lots of channels that do not stream on a 24/7 basis. We also have lots of channels that tend to go down pretty easily, due to servers' instabilities.

### Idea
To verify channels in a more intelligent way, we can proceed like this : 

- On first daily scan (Monday), 3 scans are made : one in the morning, one in the afternoon, one at night
- One JSON file of 'working streams' can be issued, with tvgid (or link and name if no data's there) and state ("working_morning", "working_afternoon", "working_night", "working_247", "blocked", "not_working")

### Principle
- Channels that are reportedly working in the morning are kept for rescanning.
- Channels that are reportedly working in the afternoon and not in the morning gets the "working_afternoon" tag, and written. Thus, channels that are reportedly working in the morning **and** in the afternoon are kept for rescanning, and channels that only work in the morning gets the "working_morning" tag, and written.
- Channels that are reportedly working in the nighttime and not in the afternoon **and** morning get the "working_night" tag, and written. Thus, channels that are reportedly working in every single instance gets the "working_247" tag, and written.
- If the server does not answer on all 3 of these instances, it gets the "not_working" tag, and written.
- If the server answers with a "403 Forbidden" status code, it gets the "blocked" tag, and written.

### Potential ameliorations
- If the server answers with a "451 Unavailable for Legal Reasons" status code (see Yandex channels for example), it gets the "geoblock" label, and written.

**Comment by Central302** [comment]
(For Note) CF97 Live (Prevously added) wiil Work only during Match that airing on WGN9.

**Comment by john412smith** [comment]
I have played with this, what I have come up with is simply check the headers only, if code is over 308 remove the channel. Requesting only the header seems to be the most efficient way. However, if the channel is geoblocked from your location, it will still flag the channel... Could be modified to account for this case and others easily... This is a base concept to work off of only.

Written in PHP as an example:

$qry = $db->query("select * from live_streams");
while($chan = $qry->fetchArray(SQLITE3_ASSOC)){
	if($headers = get_headers(base64_decode($chan['stream_url']))){
		$code = substr($headers[0], 9, 3);
		if($code >= 308){ $db->query("delete from live_streams where id =".$chan['id']); echo "."; }else{ continue;	}
	}else{
		$db->query("delete from live_streams where id =".$chan['id']);
		echo ".";
	}	
}

**Comment by freearhey** [comment]
#8145
